### PR TITLE
LibGodot with GDExtension

### PR DESCRIPTION
**Pull request by Faolan-Rad**
LibGodot is a system to allows Godot to be compiled as a library and connected into it using GDExtensions by giving a function pointer to the entry point of a GDExtension implementation before starting Godot.

LibGodot adds language support for languages that have C support but can not becompiled into a shared library.

LibGodot allows the use of different C# runtimes allowing dotnetAOT compilation, C# android support, and C# web support.

LibGodot also allows Godot to be embedded in other applications like Blender.

### CODE REVIEW

**Comment by lyuma** [comment]
Related to #68153 - this provides a means of loading the Godot Engine from a standard .NET 6 android project

**Comment by Zireael07** [comment]
Is this for 4.1 or 4.0?

**Comment by YuriSizov** [comment]
@Zireael07 You can see the milestone :) Nothing of the sort is going into 4.0 at this point. We're done with features.

**Comment by Zireael07** [comment]
Ah. Right. Mobile GitHub and my poor eyesight are to blame.

**Comment by Faolan-Rad** [comment]
It does for the most part the thing it doesn't do is allow someone using the Library to manually tell Godot to update

File: core/libgodot/libgodot.cpp [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+	initialization_function = binder;
+}
+
+void init_libgodot_resource() {
+	if (initialization_function != nullptr) {
+		Ref<GDExtension> godotlib;
+		godotlib.instantiate();
+		Error err = godotlib->initialize_extension_function(initialization_function, "LibGodot");
+		if (err != OK) {
+			ERR_PRINT("LIBGodot Had an error initialize_extension_function'");
```
Comment (fire) [inline core/libgodot/libgodot.cpp:51]: Weird grammar casing on `LIBGodot`.

**Comment by fire** [comment]
We discussed out-of-band how that @Faolan-Rad should tell us the story of his need for this feature and how this can benefit the wider community of C#, other languages and XR.

TL;DR: pending a new proposal.

**Comment by fire** [comment]
The update problem was avoided by having the libgodot user make a gdextension and have it block in the gdextension class like the "_process" method.

## Benefits

1. Does not step the MainLoop.
2. Does not handle synchronization when two different functions are called externally.

**Comment by neikeq** [comment]
Regarding C#:

No work has started yet for porting to mobile, so I can't say what will be needed. If we can't find a better solution, we would have to go with a C# project as entry-point and Godot as a shared library. However, there's actually hope that .NET 8 or 9 could allow publishing mobile projects as libraries, which would simplify things for us.

**Comment by Faolan-Rad** [comment]
I have a basic test project that I have put together in C#
[LibGodotSharp](https://github.com/RhubarbVR/LibGodotSharp)
image

**Comment by JimFFM** [comment]
@Faolan-Rad  what is the scons command to get the static lib?

**Comment by JimFFM** [comment]
@Faolan-Rad 
do you have e.g. extension_api.json  and other missing files needed to make the demo works?

**Comment by stucotso** [comment]
I'd really like to voice support for this PR.  It enables Godot to be embedded in many more scenarios, and I think would be awesome feature. I also have needs for this and was excited to see this PR in process.

In my case I have a custom application that could not be embedded 'into' Godot as it uses custom build systems, tools, languages, etc - with this approach I could import Godot as a dependency, and continue to use Godot C++ APIs for other functionality.  I'm an example of a user that Godot would gain if this type of embedding were possible.

Only minor tweak I would recommend is that the library user 'could' have the option to own the main loop, as delegating the main loop completely is not always ideal.  But even if this isn't possible I'd let the PR through and folks like myself could make minor source patches to have more main loop control.

**Comment by m4gr3d** [comment]
@Faolan-Rad For Android, Godot is already built and used as a shared library. Would this PR enables that capability for other platforms as well?

**Comment by Faolan-Rad** [comment]
In [LibGodotSharp](https://github.com/RhubarbVR/LibGodotSharp) I have set up android support

**Comment by GeorgeS2019** [comment]
FYI

#### [Is Making Advanced GUI Applications with Godot the Future?](https://medium.com/swlh/what-makes-godot-engine-great-for-advance-gui-applications-b1cfb941df3b)

Now https://github.com/RhubarbVR/LibGodotSharp works in Desktop (Win) and Android.

**Comment by Faolan-Rad** [comment]
First stable pre releases of [LibGodotSharp](https://github.com/RhubarbVR/LibGodotSharp/releases/tag/4.0.2) witch uses this system is out now.

**Comment by Faolan-Rad** [comment]
I had to fixed bugs with loading orders and API registering levels with GDextensions that made it where the project settings could not be changed before they are used I have fixed it on my [master](https://github.com/RhubarbVR/godot-lib) branch but not in this pull request because it was a problem with GDextensions.

**Comment by sferrogoo** [comment]
@Faolan-Rad I grabbed your master branch and tried to build. All worked well for static lib.  Any chance you can write a doc or example on how to use the static lib from C++ ?  I did see the C# example - but I wasn't able to easily map all the logic to C++ to try to get something to work.   If you provide info I can help to write an example to show how this works as well.  It might help folks understand value of this cool feature.

**Comment by fire** [comment]
I was puzzling over this.

Here's is the closest api doc. https://github.com/RhubarbVR/LibGodotSharp/blob/main/libgodotsharp/Native.cs

**Comment by fire** [comment]
- [ ] load dll
- [ ] object_method_bind_call
- [ ] object_method_bind_ptrcall
- [ ] object_destroy

**Comment by stucotso** [comment]
Seems it could be possible to just write c++ as standard Godot module code, not use GDExtension directly. Would be better performance and one could static link for fast compile as well.   But I’m new to Godot so maybe expert can comment?

 Thanks for this amazing pr! Can’t wait to see it happen!

**Comment by Faolan-Rad** [comment]
@stucotso 
I used GDExtension for better interrogations in to other languages and if you tried to do a shared library you would hit the linker limits on windows because of the amount of functions

**Comment by Airyzz** [comment]
Very interested in this pr! I was wondering if it enables us to render to a custom texture or OpenGL context / similar? My use case is to be able to embed the godot game view in the UI of other apps with different UI Frameworks, like Flutter.

**Comment by GeorgeS2019** [comment]
@Airyzz 
Here is [TodoList](https://github.com/RhubarbVR/LibGodotSharp/issues/4#issue-1591187688)

**Comment by GeorgeS2019** [comment]
```cs
//https://github.com/RhubarbVR/LibGodotSharp/blob/main/sharp_generator/Program.cs
[DllImport("libgodot")]
public static extern int godot_main(int argc, string[] args);

 var custom_args = new string[] { "libgodot", "--dump-extension-api", "--verbose", "--headless", "" };
if (godot_main(custom_args.Length - 1, custom_args) != 0)
 {
    throw new Exception("Godot had error");
 }
```

**Comment by migueldeicaza** [comment]
This is an amazing patch, and would allow many of us that wanted to use Godot as a library to do so, a great feature for those of us that mostly do applications, but want to sprinkle a little Godot here and there.

**Comment by GeorgeS2019** [comment]
@migueldeicaza 

We are WIP to use XAML parser to automate 3D UI using this library. Keep coming back to support this IDEA!!! we need you to push this library for 3D control in MAUI

File: core/libgodot/libgodot.h [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+#endif
+
+void libgodot_init_resource();
+
+void libgodot_project_settings_load(void *project_settings);
+
+void libgodot_scene_load(void *scene);
+
+bool libgodot_is_scene_loadable();
+
```
Comment (migueldeicaza) [inline core/libgodot/libgodot.h:58]: bool does not exist in plain C interfaces.

Comment (lyuma) [reply core/libgodot/libgodot.h:58]: Indeed, C99 requires stdbool.h to be included in order for the `bool` type to be defined.
I suggest adding this to the top of the file to permit pure C -std=c99 code to use the header:
```cpp
#ifndef __cplusplus
#include <stdbool.h>
#endif
```
I don't think there is a need to avoid use of bool because c99 support is probably widespread by now.

Comment (lyuma) [reply core/libgodot/libgodot.h:58]: I was thinking in terms of the whole godot api where bool may be used in many places...
But if it's just that one function, it might be easier to save the hassle and return int and avoid needing stdbool

File: main/main.cpp [hunk 3]
```diff
@@ -1366,11 +1370,17 @@ Error Main::setup(const char *execpath, int argc, char *argv[], bool p_second_ph
 #ifdef TOOLS_ENABLED
 		editor = false;
 #else
-		const String error_msg = "Error: Couldn't load project data at path \"" + project_path + "\". Is the .pck file missing?\nIf you've renamed the executable, the associated .pck file should also be renamed to match the executable's name (without the extension).\n";
-		OS::get_singleton()->print("%s", error_msg.utf8().get_data());
-		OS::get_singleton()->alert(error_msg);
+#ifdef LIBRARY_ENABLED
+		if (!libgodot_is_scene_loadable()) {
+#endif
+			const String error_msg = "Error: Couldn't load project data at path \"" + project_path + "\". Is the .pck file missing?\nIf you've renamed the executable, the associated .pck file should also be renamed to match the executable's name (without the extension).\n";
```
Comment (migueldeicaza) [inline main/main.cpp:1376]: Is there a reason to only enable this codepath when you are building a non-tools version?

I would love to have a single build that does both an editor and the library

**Comment by fire** [comment]
@Faolan-Rad do you remember why in your fork https://github.com/RhubarbVR/godot-lib/actions/runs/4277032243/jobs/7445556782 is failing because of the use of the new API levels server and scene?

I took it out of my build to attempt to get tests to pass.

**Comment by Faolan-Rad** [comment]
I have updated this to align with the latest 'main' branch. Additionally, I've incorporated a C# entry point. The bug fixes related to project settings loading have been extracted and will be presented in a separate pull request at some point.

File: modules/mono/editor/Godot.NET.Sdk/Godot.SourceGenerators/GodotPluginsInitializerGenerator.cs [hunk 4]
```diff
@@ -24,10 +24,13 @@ public void Execute(GeneratorExecutionContext context)
 
 namespace GodotPlugins.Game
 {
-    internal static partial class Main
+    public static partial class Main
     {
+        /// <summary>
+        /// Should not be touched endless you know what your doing this is a UnmanagedCallersOnly function
```
Comment (kelteseth) [inline modules/mono/editor/Godot.NET.Sdk/Godot.SourceGenerators/GodotPluginsInitializerGenerator.cs:30]: **unless** you know know what your doing

Comment (lyuma) [reply modules/mono/editor/Godot.NET.Sdk/Godot.SourceGenerators/GodotPluginsInitializerGenerator.cs:30]: if we're going to have a grammar fix, might as well go the whole way:

**unless** you know know what **you are** (or **you're**, not ~~your~~) doing **. (period)** This is is a UnmanagedCallersOnly function
```suggestion
        /// Should not be touched unless you know what you are doing. This is a UnmanagedCallersOnly function.
```

File: core/extension/gdextension_interface.cpp [hunk 5]
```diff
@@ -1077,10 +1077,12 @@ static GDExtensionMethodBindPtr gdextension_classdb_get_method_bind(GDExtensionC
 		return nullptr;
 	}
 	ERR_FAIL_COND_V(!mb, nullptr);
+#ifndef LIBRARY_ENABLED
 	if (mb->get_hash() != p_hash) {
 		ERR_PRINT("Hash mismatch for method '" + classname + "." + methodname + "'.");
 		return nullptr;
 	}
+#endif
```
Comment (dsnopek) [inline core/extension/gdextension_interface.cpp:1085]: Why do we want to disable checking for hash mismatches here? Given that GDExtension does ptrcalls to any methods returned, calling a method with an unexpected signature can lead to crashes.

File: core/extension/gdextension.cpp [hunk 6]
```diff
@@ -495,17 +503,21 @@ void GDExtension::close_library() {
 	library = nullptr;
 }
 
+bool GDExtension::is_embedded() const {
+	return embedded;
+}
+
 bool GDExtension::is_library_open() const {
-	return library != nullptr;
+	return (library != nullptr) || embedded;
```
Comment (dsnopek) [inline core/extension/gdextension.cpp:511]: Having to check for `embedded` at runtime in so many places doesn't feel quite right. Compiling Godot in this "LibGodot mode" is not the common case. Can all these checks be `#ifndef`'d away for when running Godot normally?

**Review by dsnopek — COMMENTED** [review]
In general, I like the idea of having the option to compile Godot as a library!

However, I feel like the way the changes to GDExtension are done is kind of messy. I'd like this a lot better if the GDExtension changes could be more self-contained, rather than spread out in runtime `if (embedded)` checks all over the place. The code this changes is only going to get more complicated (for example, after PR #80284), and as it evolves, all contributors that work on it would need to know how to do the right checks so it continues to work in both modes.

Could there maybe be a `GDExtensionEmbedded`  class that's a child class of `GDExtension` so that `GDExtension` could be left unchanged? Or, could the GDExtension interface (ie the API defined by `gdextension_interface.h`) be used through another mechanism without going through the `GDExtension` and `GDExtensionManager` classes? Or, some other way that all the changes for this embedded mode would live separately from the code used when running Godot normally...

**Comment by Faolan-Rad** [comment]
@dsnopek  I do like your suggestion I would like to do it with the a GDExtensionBase and then there would be the normal GDExtension and EmbeddedGDExtension, and the manager itreacts with the base class I would also like to make it where it is not just binding one EmbeddedGDExtension you can have multiple. I would also like to make it assigned able from other GDExtension this would be more for having a GDExtension for lua implementation and then registering the lua code as a GDExtension

Currently, I am prioritizing something else for my game specifically, working on the _mix method for AudioStreamPlayback in C# [PR](https://github.com/godotengine/godot/pull/81004) so I can efficiently implement opus for my game. I do want to come back to this soon but it might be a second.

**Comment by dsnopek** [comment]
> Or, some other way that all the changes for this embedded mode would live separately from the code used when running Godot normally...

Another possible idea: What if instead of `GDExtension` directly using `OS::open_dynamic_library()` and `OS::get_dynamic_symbol_handle()`, it used a new layer (ex `GDExtensionLoader`?) that could have its implementation replaced when in "LibGodot mode" (maybe via a `GDExtensionLoaderEmbedded` child class)? This could be used to prevent really loading dynamic libraries, and instead look up the entry point function pointers from a list. I think this would require less code changes than making a `GDExtensionEmbedded` child class, and it should still be easy to `#ifndef` away the LibGodot version of the "GDExtension loader layer" when not needed.

**Comment by dannflor** [comment]
2 Questions:
1. Would LibGodot unlock iOS .NET exports as well? I see lots of discussion about Android everywhere but precious few mentioning iOS.
2. Conditional on 1, would you be interested in accepting help on this branch implementing the prior suggestions? The lack of iOS exports is a real painful blocker for me right now and I'm eager to push work forward on it any way possible.

**Comment by Faolan-Rad** [comment]
I haven't directly tested it but because I believe ios requires Aot there might be problems because godot sharp uses reflection so that would complicate things  so throughs would have to be fixed but if you look at it from if you could run the jit on iOS it would work fine from what I know

**Comment by Faolan-Rad** [comment]
@dannflor [iOS c# docs](https://learn.microsoft.com/en-us/xamarin/ios/internals/limitations)
Would need to work around the limitations listed there for the iOS build but this would be the first step closer to it then it is fixing the godot sharp to make it Aot compatible

**Comment by migueldeicaza** [comment]
This works great with the SwiftGodot bindings.  

https://github.com/migueldeicaza/SwiftGodotKit

**Comment by dannflor** [comment]
> [iOS c# docs](https://learn.microsoft.com/en-us/xamarin/ios/internals/limitations) Would need to work around the limitations listed there for the iOS build
I took glance at the limitations and it seems like they shouldn't be a problem. I certainly don't use any of those features in my scripting, though I'm not sure if the C# engine code uses any of it. Are you aware of any Godot code we'd have to patch around?

> This works great with the SwiftGodot bindings. https://github.com/migueldeicaza/SwiftGodotKit
Have you used it on iOS? The examples in the repo are still targeting macOS. And I'm a little confused how I would glue the xcframework of libgodot to a full godot project I have (not written in SwiftGodot, just plain ol c#) and get a running game.

**Comment by mohumohu-corp** [comment]
> @dannflor [iOS c# docs](https://learn.microsoft.com/en-us/xamarin/ios/internals/limitations) Would need to work around the limitations listed there for the iOS build but this would be the first step closer to it then it is fixing the godot sharp to make it Aot compatible

Does the Mono interpreter solve some of the problems with the above link?
https://learn.microsoft.com/en-us/dotnet/maui/macios/interpreter

**Comment by kisg** [comment]
This PR is very important for many different use cases. We have a similar in-house implementation that we would be happy to contribute to this PR as patches, if there is interest. I wrote more about it here: https://github.com/godotengine/godot-proposals/issues/6267#issuecomment-1721686318

**Comment by Faolan-Rad** [comment]
Okay so some one correct me if i am wrong Godot could still be an excusable that can be used as a shared library if the symbols that are needed are exported so at that rate it would be able to share a single build of Godot for each platform and they would all have the shared library ability.

**Comment by Faolan-Rad** [comment]
So some things that could be changed is making it not add code to GDExtension and have it work at the os shared library level. Also it might be considered to maybe to just make the normal build export the 2 needed symbols the main function and the binding function

**Comment by Faolan-Rad** [comment]
I am going to go work on the things I mentioned above

**Comment by GeorgeS2019** [comment]
@kattkieru

Please share the exact link to the PR

**Comment by kisg** [comment]
@GeorgeS2019

There is no PR yet, I showed it at GodotCon after the talk of @migueldeicaza, and a small demo is here: https://x.com/JoshBromberg1/status/1721564103859114132

The current version supports embedding Godot into iOS apps and also render multiple Godot windows into surfaces (OpenGL / Vulkan) provided by the host application. It also includes input support. Other platforms will also be supported.

I discussed it with @akien-mga at GodotCon, we will start the review process after the release of Godot 4.2.

**Comment by kattkieru** [comment]
Ahh I was under the impression it was already up. I’d love to see more, especially the python bindings part. I’m very interested in embedding the Godot viewport in other applications.

Please do link here when you’ve uploaded the pull.

**Comment by haltingstate** [comment]
Can you fix the conflicts for this, so it can be merged in?

@Faolan-Rad

**Comment by zoubirabde9a** [comment]
When can we start using this

@Faolan-Rad

**Comment by haltingstate** [comment]
> Is this slated for 4.1 or 4.0?

This needs to be merged in for 4.2

**Comment by choi303** [comment]
Finish this already @Faolan-Rad

**Comment by AThousandShips** [comment]
> Can you fix the conflicts for this, so it can be merged in?

It'd need to be approved before merging it, but fixing conflicts would improve the workflow

> This needs to be merged in for 4.2

It won't be, we are in feature freeze and won't merge features for 4.2

**Comment by AThousandShips** [comment]
@choi303 Please do not rush people or make such demands, please be polite and constructive

**Comment by choi303** [comment]
> @choi303 Please do not rush people or make such demands, please be polite and constructive

Okay sorry

**Comment by haltingstate** [comment]
> > Can you fix the conflicts for this, so it can be merged in?
> 
> It'd need to be approved before merging it, but fixing conflicts would improve the workflow
> 
> > This needs to be merged in for 4.2
> 
> It won't be, we are in feature freeze and won't merge features for 4.2

Can I sponsor a developer to finish this, for 4.2 rc 2?

Who do I talk to?

We need it, to port our command line CI/CD pipeline from Unity to Godot.

**Comment by haltingstate** [comment]
> > Can I sponsor a developer to finish this, for 4.2 rc 2?
> 
> No, it won't be in 4.2, as I explained, we are in feature freeze and will not add any features to 4.2 before release

Ok, but what about release candidate 4.2.1.

I will make a blog post and I will make some document about doing command line CI and unit test in Godot 4.2 / 4.3 and maybe Spaces integration for testing pipeline, if someone can help with this. And will create github with code examples and can put a developer on updating the godot website documentation for usage examples.

**Comment by AThousandShips** [comment]
We normally do not backport features, it would depend on the result

This is a popular idea, and will hopefully be reviewed for 4.3 and might be merged for then

But please, understand that the process of development takes time, and there's no rushing these things, right now there are conflicts that needs to be resolved, and reviews to be made, when people have the time to focus on this

So let's wait and see when this gets feedback and how it turns out, no need to continue this back and forth here, it won't speed anything up, once 4.2 is released this can get a review if the people necessary for reviewing this are able to 🙂

**Comment by nukeop** [comment]
Where can the feature freeze schedule be found?

**Comment by AThousandShips** [comment]
[Here](https://godotengine.org/article/release-management-4-1/), now please let us stop with off topic discussions and leave this for when something updates 🙂

**Comment by nukeop** [comment]
I don't see any conrete dates or even estimates in that blog post. Is there any up to date schedule?

**Comment by AThousandShips** [comment]
It states the lengths of the cycles, we don't maintain an official schedule, but we maintain the cycles

Any further discussions about release management will be taken elsewhere or be considered spam, thank you and have a nice day

**Comment by kisg** [comment]
@haltingstate 

> 
> Can I sponsor a developer to finish this, for 4.2 rc 2?
> 
> Who do I talk to?
> 
> We need it, to port our command line CI/CD pipeline from Unity to Godot.

As I wrote above last week (and showed it at GodotCon), we at [Migeran](https://migeran.com) have a much-improved version of this solution that allows the embedding of Godot's UI in the host process as well.

We plan to go through the upstreaming review process during the 4.3 development cycle (as discussed with the Godot maintainers). If you have specific requirements or need custom builds for 4.2 or other commercial support of this feature, you may reach out to me directly.

**Comment by haltingstate** [comment]
> @haltingstate
> 
> > Can I sponsor a developer to finish this, for 4.2 rc 2?
> > Who do I talk to?
> > We need it, to port our command line CI/CD pipeline from Unity to Godot.
> 
> As I wrote above last week (and showed it at GodotCon), we at [Migeran](https://migeran.com) have a much-improved version of this solution that allows the embedding of Godot's UI in the host process as well.
> 
> We plan to go through the upstreaming review process during the 4.3 development cycle (as discussed with the Godot maintainers). If you have specific requirements or need custom builds for 4.2 or other commercial support of this feature, you may reach out to me directly.

Thank you. That is great.

Can you link to the github repo?

**Comment by nukeop** [comment]
> This is a popular idea, and will hopefully be reviewed for 4.3 and might be merged for then

Seeing as there are now previews for 4.3, I'm guessing it hasn't?

**Comment by Faolan-Rad** [comment]
@adamscott  if you would like to chat about this i am free anytime you can email me at [faolan@rhubarbvr.net](mailto:faolan@rhubarbvr.net) or contact me on discord with faolanrad

**Comment by kisg** [comment]
We also have a new version almost ready for review.

It works like this:
* adds a single public symbol to the GDExtension API:
```
GDExtensionObjectPtr gdextension_create_godot_instance(int p_argc, char *p_argv[], GDExtensionInitializationFunction p_init_func);
```
* this function initializes the Godot instance, the GDExtension API and returns a GDExtension object of type GodotInstance
* The GodotInstance object can be used to start, iterate, shutdown the Godot instance.
* It is also possible to pass a native surface handle to the start() call so the Godot UI can be embedded into the host process.
  * This is currently implemented for Apple systems (MacOS and iOS), with a design that will make it feasible to support Windows and Linux as well.
  * It is also possible to pass a native surface handle to a Godot Window, which allows us to move the contents of a Godot Window to a separate host surface.

Because we use the GDExtension API, language bindings will require only minimal changes. We will provide patches for SwiftGodot and godot-cpp initially.

We have a few finishing touches left to do next week before we can submit the PR.

**Comment by ppphp** [comment]
I have been using the branch for quite some time, make godot app as a Qt widget

**Comment by GeorgeS2019** [comment]
> We plan to submit the PR by the end of next week, we just have a few finishing touches left.

# [Godot4-dotnet](https://github.com/raulsntos/godot-dotnet)

**Comment by GeorgeS2019** [comment]
@Faolan-Rad 

What do we need to do [here ](https://github.com/raulsntos/godot-dotnet/pull/3)or to implement **to achieve** Auto Generate c# Bindings for e.g. c++ GDExtension e.g.[ GDMP](https://github.com/j20001970/GDMP/issues/25)

**Comment by kisg** [comment]
FYI: I just created our LibGodot PR: https://github.com/godotengine/godot/pull/90510
We also created a preconfigured build environment with sample apps for easy testing: https://github.com/migeran/libgodot_project

**Comment by GeorgeS2019** [comment]
# [WIP] [Windows platform: **libgodot.dll**](https://github.com/JiepengTan/godot/tree/libgodot_windows_build)

The libgodot.dll comes from 
- library_type=shared_library

`scons platform=windows target=template_debug arch=x86_64 verbose=yes msvc=yes library_type=shared_library dev_build=yes debug_symbols=yes`

bin\ **godot.windows.template_debug.dev.x86_64.dll**

`scons platform=windows arch=x86_64 verbose=yes msvc=yes dev_build=yes debug_symbols=yes`

bin\ **godot.windows.editor.dev.x86_64.exe** --dump-extension-api

With this, provide the **json** and the **dll** to the [godot-dotnet](https://github.com/raulsntos/godot-dotnet) project.
1. bin\ **extension_api.json**
2. bin\ **godot.windows.template_debug.dev.x86_64.dll**

The output will be c# wrappers for the **godot.windows.template_debug.dev.x86_64.dll**

Users can then write c# applications to **embed** godot as a library in .NET applications

Very similar to: [LibGodotSharp](https://github.com/RhubarbVR/LibGodotSharp)

**Comment by Repiteo** [comment]
Superseded by #110863
